workers/unity/Packages/io.improbable.worker.sdk/EventTracing.cs

```diff
@@ -0,0 +1,403 @@
+
+                    var newSpan = newItem.Span.GetValueOrDefault();
+                    ApiInterop.Memcpy(newSpan.Id.Data, itemContainer->ItemUnion.Span.Id.Data, SpanId.SpanIdSize);
+
+                    newSpan.CauseCount = (int) itemContainer->ItemUnion.Span.CauseCount;
+                    for (var i = 0; i < newSpan.CauseCount; i++)
+                    {
+                        newSpan.Causes[i] = new SpanId();
+                        fixed (byte* spanIdDest = newSpan.Causes[i].Data)
+                        {
+                            ApiInterop.Memcpy(spanIdDest, itemContainer->ItemUnion.Span.Causes[i].Data, SpanId.SpanIdSize);
+                        }
+                    }
+
+                    break;
+                case ItemType.Event:
+                    newItem.Event = new Event();
+
+                    var newEvent = newItem.Event.GetValueOrDefault();
+                    newEvent.Id = new SpanId();
+                    ApiInterop.Memcpy(newEvent.Id.Data, itemContainer->ItemUnion.Event.Id.Data, SpanId.SpanIdSize);
+
+                    newEvent.UnixTimestampMillis = itemContainer->ItemUnion.Event.UnixTimestampMillis;
+                    newEvent.Type = ApiInterop.FromUtf8Cstr(itemContainer->ItemUnion.Event.Type);
+                    newEvent.Message = ApiInterop.FromUtf8Cstr(itemContainer->ItemUnion.Event.Message);
+
+                    newEvent.Data = Marshal.PtrToStructure<TraceEventData>(itemContainer->ItemUnion.Event.Data);
+                    break;
+                default:
+                    throw new NotSupportedException("Invalid Item Type provided.");
+            }
+
+            return newItem;
+        }
+
+        internal static unsafe CEventTrace.Item* ConvertItem(IOStorage storage, Item item)
+        {
+            var newItem = CEventTrace.ItemCreate(storage.Storage, null);
+            newItem->ItemUnion = new CEventTrace.Item.Union();
+            newItem->ItemType = (CEventTrace.ItemType) item.ItemType;
+
+            switch (item.ItemType)
+            {
+                case ItemType.Span:
+                    var spanItem = item.Span.GetValueOrDefault();
+                    newItem->ItemUnion.Span = new CEventTrace.Span();
+                    newItem->ItemUnion.Span.Id = ParameterConversion.ConvertSpanId(spanItem.Id);
+                    newItem->ItemUnion.Span.CauseCount = (uint) spanItem.CauseCount;
+                    for (var i = 0; i < newItem->ItemUnion.Span.CauseCount; i++)
+                    {
+                        newItem->ItemUnion.Span.Causes[i] = ParameterConversion.ConvertSpanId(spanItem.Causes[i]);
+                    }
+
+                    break;
+                case ItemType.Event:
+                    var eventItem = item.Event.GetValueOrDefault();
+                    ParameterConversion.ConvertEvent(eventItem, internalEvent =>
+                    {
+                        newItem->ItemUnion.Event = internalEvent;
+                    });
```

Comment:
This is still a little unsafe. The closure passed to `ConvertEvent` needs to encapsulate the entire process until the raw event is consumed. 

As it stands, the pointers inside `newItem->ItemUnion.Event` could be invalid after we exit this closure. We then return the item to the call who presumably does something with it.

Reply:
Discussed offline, refactored this approach to control the lifetime of the `internalEvent` until it's consumed by using a closure.